yt/yt/core/concurrency/action_queue.cpp

```diff
@@ -678,4 +680,51 @@ IInvokerPtr CreateCodicilGuardedInvoker(IInvokerPtr underlyingInvoker, TString c
 
 ////////////////////////////////////////////////////////////////////////////////
 
+class TWatchdogInvoker
+    : public TInvokerWrapper
+{
+public:
+    TWatchdogInvoker(
+        IInvokerPtr invoker,
+        const NLogging::TLogger& logger,
+        TDuration threshold)
+        : TInvokerWrapper(std::move(invoker))
+        , Logger(logger)
+        , Threshold_(DurationToCpuDuration(threshold))
+    { }
+
+    void Invoke(TClosure callback) override
+    {
+        UnderlyingInvoker_->Invoke(BIND_NO_PROPAGATE(
+            &TWatchdogInvoker::RunCallback,
+            MakeStrong(this),
+            Passed(std::move(callback))));
+    }
+
+private:
+    NLogging::TLogger Logger;
+    TCpuDuration Threshold_;
+
+    void RunCallback(TClosure callback)
+    {
+        TCurrentInvokerGuard currentInvokerGuard(this);
+        TFiberSliceTimer fiberSliceTimer(Threshold_, [&, this] (TCpuDuration execution) {
+            YT_LOG_WARNING("Callback executed for too long without interruptions (Callback: %v, Execution: %v)",
+                callback.GetHandle(),
```

Comment:
For diagnostics it would be better to use Location from TBindStateBase if it is enabled.